File: Tutorials/Scala/5 Topics/producer/src/main/scala/ru/example/rmq/topics/Producer.scala
package ru.example.rmq.topics

import com.rabbitmq.client._
import java.nio.charset.StandardCharsets
import scala.util.Using

object Producer {
  def main(args: Array[String]): Unit = {
    val EXCHANGE_NAME = "topic_logs"
    val virtualHost   = "/"

    if (args.length < 3) {
      System.out.println("Usage: Producer hostname username password [routingKey] [message...]")
      System.exit(-1)
    }

    val hostname = args(0)
    val username = args(1)
    val password = args(2)

    val routingKey = if (args.length > 3) args(3) else "anonymous.info"
    val message    = if (args.length > 4) args.slice(4, args.length).mkString(" ") else "Hello World!"

    val factory = new ConnectionFactory
    factory.setHost(hostname)
    factory.setUsername(username)
    factory.setPassword(password)
    factory.setVirtualHost(virtualHost)

    Using.Manager { use =>
      try {
        val connection = use(factory.newConnection)
        val channel    = use(connection.createChannel)

        channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.TOPIC)
        channel.basicPublish(
          EXCHANGE_NAME,
          routingKey,
          null,
          message.getBytes(StandardCharsets.UTF_8)
        )
        println(s"[x] Sent '$routingKey': '$message'")
      } catch {
        case e: Exception =>
          println(e.getLocalizedMessage)
          sys.exit(-1)
      }
    }
    sys.exit(0)
  }
}

File: Tutorials/Java/5 Topics/Producer/src/main/java/Producer.java
import com.rabbitmq.client.BuiltinExchangeType;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class Producer {
    public static void main(String[] argv) throws Exception {
        String hostname;
        String username;
        String password;
        String virtualHost = "/";
        String EXCHANGE_NAME = "topic_logs";
        String routingKey;
        String message;

        if (argv.length < 3) {
            System.out.println("Usage: Producer hostname username password [routingKey] [message...]");
            System.exit(-1);
        }

        hostname = argv[0];
        username = argv[1];
        password = argv[2];

        if (argv.length > 3) routingKey = argv[3];
        else routingKey = "anonymous.info";

        if (argv.length > 4) message = String.join(" ", Arrays.copyOfRange(argv, 4, argv.length));
        else message = "Hello World!";

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(hostname);
        factory.setUsername(username);
        factory.setPassword(password);
        factory.setVirtualHost(virtualHost);

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {
            channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.TOPIC);
            channel.basicPublish(EXCHANGE_NAME, routingKey, null, message.getBytes(StandardCharsets.UTF_8));
            System.out.println(" [x] Sent '" + routingKey + "':'" + message + "'");
        }
    }
}
